### ISSUE TITLE
kubelet won't start on GCE VM without a metadata server.

### ISSUE THREAD

**Issue by oc243** [kubernetes/kubernetes#30195]
If you start kubelet on a GCE VM with `local-up-cluster.sh`, kubelet fails to come up. The logs show that it is sat spinning on
`I0807 17:46:01.660729   20991 metadata.go:205] Failed to Get service accounts from gce metadata server: Get http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/: dial tcp: lookup metadata.google.internal.: no such host`.

This looks like a bug introduced in ea1a459 and affects v1.3.4. A comment in this commit says "Given that we are on GCE we should keep retrying until the metadata server responds.", however there is no guarantee that by using GCE you will have a metadata server.

**Comment by yujuhong** [comment]
@oc243, my understanding is that very gce instance has a metadata server. 
From https://cloud.google.com/compute/docs/storing-retrieving-metadata#querying,
"Every instance stores its metadata on a metadata server. "

Also from https://cloud.google.com/compute/docs/networking:
"Each instance has a metadata server that also acts as a DNS resolver"

In what case that your GCE instances doesn't have access to the metadata server?

**Comment by dchen1107** [comment]
cc/ @mikedanese Could you please help with this one?

**Comment by oc243** [comment]
Ah, I actually am not using Google's DNS resolver, which is probably why the lookup is failing…

**Comment by yujuhong** [comment]
@oc243 did you setup a network proxy for your cluster?

**Comment by oc243** [comment]
No, but I suspect the custom DNS is breaking the google cloud VM has a metadata server assumption?

**Comment by mikedanese** [comment]
Is this solved then?

**Comment by oc243** [comment]
I'd say not. Currently if you run on GCE you need to use google's DNS resolver to resolve `metadata.google.internal`. If you use a custom resolver and so can't resolve this domain then it should backoff rather than spinning for this domain to become available, I think.

**Comment by yujuhong** [comment]
@oc243 I agree it's bad that kubelet won't start when you have custom DNS. If you disable the check by patching kubelet, does the rest of the kubernetes work properly with the custom DNS?

**Comment by mikedanese** [comment]
Can you add GCE nameserver as the last entry in your resolv.conf or add a record in the nameserver you are running for the metadata server?

**Comment by mikedanese** [comment]
As a workaround. This should be fairly easy to fix
